### PR TITLE
Conform to `ExpressibleByArrayLiteral` and `ExpressibleByDictionaryLiteral`

### LINKED ISSUE #48: `ExpressibleByArrayLiteral` and `ExpressibleByDictionaryLiteral` conformances
It would be _very_ convenient to have these conformances, since non-empty collections could be initialized as naturally as the built-in collection types:

```swift
let x: NonEmpty<[Int]> = [1, 2, 3]
let y: NonEmpty<[Int: String]> = [1: "abc", 2: "def"]
```

However, since `ExpressibleByArrayLiteral` and `ExpressibleByDictionaryLiteral` don't allow their initializers to fail, we would have to crash at runtime, which might not be desirable given the description of the library as a compile-time guarantee:

```swift
init(arrayLiteral: Element...) {
  guard !arrayLiteral.isEmpty else { fatalError("Non-empty array literal!") }
  self.init(rawValue: arrayLiteral)!
}
```

or

```swift
init(arrayLiteral: Element...) {
  assert(!arrayLiteral.isEmpty, "Non-empty array literal!")
  self.init(rawValue: arrayLiteral)!
}
```

Thoughts?

### PR DESCRIPTION
**Pull request by fwcd**
### Fixes #48 

This branch adds experimental support for the aforementioned protocols.

### CODE REVIEW

**Comment by stephencelis** [comment]
@fwcd Check out how we implemented it in swift-tagged: https://github.com/pointfreeco/swift-tagged/blob/f3f773a5e13f3c8f0ab1ce2ae6378058acefa87d/Sources/Tagged/Tagged.swift#L289-L314

I think you could do similar here to conditionally support any underlying collection that conforms to the `ExpressibleBy*` protocol.

**Comment by fwcd** [comment]
That's a neat workaround to the issue that Swift doesn't have a spread operator yet and definitely the simpler solution.

Unfortunately, there is another issue, certain overloads now become ambigous with `[3, 4, 5]` being interpretable as both an array and a `NonEmpty`:

```
.../git/swift-nonempty/Tests/NonEmptyTests/NonEmptyTests.swift:124:20: error: ambiguous use of 'isDisjoint(with:)'
    XCTAssertFalse(NonEmptySet(1, 2, 3).isDisjoint(with: [3, 4, 5]))
                   ^
.../git/swift-nonempty/Sources/NonEmpty/NonEmpty+SetAlgebra.swift:75:15: note: found this candidate
  public func isDisjoint(with other: NonEmpty) -> Bool {
              ^
.../git/swift-nonempty/Sources/NonEmpty/NonEmpty+SetAlgebra.swift:79:15: note: found this candidate
  public func isDisjoint(with other: Collection) -> Bool {
              ^
```

**Comment by stephencelis** [comment]
> Unfortunately, there is another issue, certain overloads now become ambigous with `[3, 4, 5]` being interpretable as both an array and a `NonEmpty`:

Hm! Could you try adding a `@_disfavoredOverload` annotation to the one that takes a `NonEmpty`?

**Comment by fwcd** [comment]
That did the trick, thanks! The test suite now passes.

**Review by stephencelis — APPROVED** [review]
Awesome! Thanks for taking the time to contribute :smile:

File: Sources/NonEmpty/NonEmpty+ExpressibleByArrayLiteral.swift [hunk 1]
```diff
@@ -0,0 +1,10 @@
+extension NonEmpty: ExpressibleByArrayLiteral where Collection: ExpressibleByArrayLiteral {
+  public init(arrayLiteral elements: Element...) {
+    assert(!elements.isEmpty, "Cannot initialize a NonEmpty array from an empty literal!")
```
Comment (stephencelis) [inline Sources/NonEmpty/NonEmpty+ExpressibleByArrayLiteral.swift:3]: Just finessing the wording and switching to be a precondition.

```suggestion
    precondition(!elements.isEmpty, "Can't construct \(Self.self) from empty array literal")
```

File: Sources/NonEmpty/NonEmpty+ExpressibleByDictionaryLiteral.swift [hunk 2]
```diff
@@ -0,0 +1,10 @@
+extension NonEmpty: ExpressibleByDictionaryLiteral where Collection: ExpressibleByDictionaryLiteral {
+  public init(dictionaryLiteral elements: (Collection.Key, Collection.Value)...) {
+    assert(!elements.isEmpty, "Cannot initialize a NonEmpty dictionary from an empty literal!")
```
Comment (stephencelis) [inline Sources/NonEmpty/NonEmpty+ExpressibleByDictionaryLiteral.swift:3]: More wording:

```suggestion
    precondition(!elements.isEmpty, "Can't construct \(Self.self) from empty dictionary literal")
```

**Pull request merged**